scikit-talk/turndynamics.py

```diff
@@ -0,0 +1,286 @@
+          return pd.NA
+      else:
+          h,m,s = text.split(':')
+          return int(datetime.timedelta(hours=int(h),
+                                        minutes=int(m),
+                                        seconds=float(s)).total_seconds()*1000)
+
+  # number of unique sources
+  def _getsourceindex(source):
+      return n_sources.index(source) + 1
+
+  # talk, laugh, breath, or other conduct classification
+  def _getnature(utterance):
+      if pd.isna(utterance) == True:
+        return pd.NA
+      if utterance == '[laugh]':
+        return 'laugh'
+      if utterance == '[breath]':
+        return 'breath'
+      if utterance in ['[cough]', '[sneeze]', '[nod]', '[blow]', '[sigh]',
+                        '[yawn]', '[sniff]', '[clearsthroat]',
+                        '[lipsmack]', '[inhales]', '[groan]']:
+        return utterance
+      else:
+        return 'talk'
+
+  # count number of characters
+  def _getnchar(utterance):
+      if pd.isna(utterance) == True:
+        return pd.NA
+      else:
+        utterance = Counter(utterance.replace(" ",""))
+        return sum(utterance.values())
+
+  # create a 'window' for each utterance
+  # The window looks at 10s prior the begin of the current utterance (lookback)
+  # Only turns that begin within this lookback are included
+  # in the window. This means that if the prior turn began later
+  # than 10s before the current utterance, then the prior turn is
+  # not included in the window.
+  def _createwindow(begin, participant):
+    lookback = 10000
+    lookfwd = 0
+    filter = (df_transitions['begin'] >= (begin - lookback)) & (df_transitions['begin'] <= (begin + lookfwd))
+    window = df_transitions.loc[filter]
+    # identify who produced the utterance
+    window['turnby'] = np.where(window['participant'] == participant, 'self',
+                                'other')
+    # calculate duration of all turns in window
+    stretch = window['end'].max() - window['begin'].min()
+    # calculate sum of all turn durations
+    talk_all = window['duration'].sum()
+    # calculate amount of talk produced by the participant in relation
+    # to the total amount of talk in the window
+    try:
+      talk_rel = window.loc[window['turnby'] == 'self']['duration'].sum() / talk_all
+    except ZeroDivisionError:
+      talk_rel = pd.NA
+    # calculate amount of loading of the channel
+    # (1 = no empty space > overlap, < silences)
+    load = talk_all / stretch
+    # calculate total amount of turns in this time window
+    turns_all = len(window.index)
+    # calculate amount of turns by this participant relative to turns by others
+    try:
+      turns_rel = (len(window[window['turnby'] == 'self'].index)) / turns_all
+    except ZeroDivisionError:
+      turns_rel = pd.NA
+
+    participants = window['participant'].nunique()
+    # create list of all measures computed
+    measures = [talk_all, talk_rel, load, turns_all, turns_rel, participants]
+    return measures
```

Comment:
This is a first (untested) python port of my R 'windowed transitions' code. As it needs to run on every single row in the db (creating a 10s window and computing these summary measures for it) it's likely to be quite a drag on performance — optimization and modularization needed.

Reply:
[also just testing the review mechanics — don't mind me if I'm doing this wrong)